### PR TITLE
Adds BufferVec

### PR DESCRIPTION
**Pull request by breeswish**
Signed-off-by: Breezewish <breezewish@pingcap.com>

## What have you changed? (mandatory)

`BufferVec` is a data structure for efficiently holding multiple binary strings (like `Vec<Vec<u8>>`). Each binary string can be in different length.

`BufferVec` supports random read and sequential write. `BufferVec` does not support random write.

`BufferVec` will be used to support `LazyBatchColumn`, as well as `VectorValue::Bytes` in future may be.

## What are the type of the changes? (mandatory)

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

New unit tests.

### CODE REVIEW

**Comment by breeswish** [comment]
It does not change existing code, so no integration test is needed.

**Comment by breeswish** [comment]
@sticnarf @lonng @rleungx  PTAL, thanks!

File: src/coprocessor/codec/batch/buffer_vec.rs [hunk 1]
```diff
@@ -0,0 +1,800 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+#![allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (breeswish) [reply src/coprocessor/codec/batch/buffer_vec.rs]: it is not used elsewhere. In the next PR it shall be removed.

File: src/coprocessor/codec/batch/buffer_vec.rs [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+        assert!(!v.is_empty());
+        assert_eq!(v[0], [0xCA, 0xB]);
+        assert_eq!(format!("{:?}", v), "[CA0B]");
+
+        v.shift(1);
+        assert_eq!(v.len(), 0);
+        assert_eq!(v.total_len(), 0);
+        assert!(v.is_empty());
+        assert_eq!(format!("{:?}", v), "[]");
+        assert!(v.is_empty());
```
Comment (rleungx) [inline src/coprocessor/codec/batch/buffer_vec.rs]: ditto

File: src/coprocessor/codec/batch/buffer_vec.rs [hunk 3]
```diff
@@ -0,0 +1,800 @@
...
+        assert_eq!(v.total_len(), 0);
+        assert!(v.is_empty());
+        assert_eq!(format!("{:?}", v), "[]");
+
+        v.truncate(1);
+        assert_eq!(v.len(), 0);
+        assert_eq!(v.total_len(), 0);
+        assert!(v.is_empty());
+        assert_eq!(format!("{:?}", v), "[]");
+        assert!(v.is_empty());
```
Comment (rleungx) [inline src/coprocessor/codec/batch/buffer_vec.rs]: Why do we need to assert two times?

File: src/coprocessor/codec/batch/buffer_vec.rs [hunk 4]
```diff
@@ -0,0 +1,800 @@
...
+        v.pop();
+        assert_eq!(format!("{:?}", v), "[]");
+        assert!(v.is_empty());
+
+        v.shift(0);
+        assert_eq!(format!("{:?}", v), "[]");
+        assert!(v.is_empty());
+
+        v.push(&[0xAA, 0x0, 0xB]);
+        assert_eq!(v.len(), 1);
```
Comment (rleungx) [inline src/coprocessor/codec/batch/buffer_vec.rs]: Can we abstract a function to do these assertions?

Comment (breeswish) [reply src/coprocessor/codec/batch/buffer_vec.rs]: The abstracted function may be more complicated than the thing it is testing.

**Comment by breeswish** [comment]
@kennytm There should be not much difference talking about performance. The main benefit is that BufferVec is more controllable that can be easily modified to use different allocators in the next step.

File: src/coprocessor/codec/batch/buffer_vec.rs [hunk 5]
```diff
@@ -0,0 +1,785 @@
...
+                            self.data.as_mut_ptr().add(data_write_offset),
+                            write_len,
+                        );
+                        self.offsets[offsets_write_offset] = data_write_offset;
+                        offsets_write_offset += 1;
+                        data_write_offset += write_len;
+                    }
+                }
+                // The last buffer does not have an ending offset
+                if retain_arr[len - 1] {
```
Comment (lonng) [inline src/coprocessor/codec/batch/buffer_vec.rs]: L220-L232 can merge to L208 with an if expression, e.g: `let write_len = if i == len-1 {self.data.len() - self.offsets[i]} else {self.offsets[i + 1] - self.offsets[i]};`

Comment (breeswish) [reply src/coprocessor/codec/batch/buffer_vec.rs]: This if expression will be checked every time even for `0 .. len - 1`, which should be less effective than current impl.

Comment (breeswish) [reply src/coprocessor/codec/batch/buffer_vec.rs]: This if expression will be checked every time even for `0 .. len - 1`, which should be less effective than current impl.

Comment (lonng) [reply src/coprocessor/codec/batch/buffer_vec.rs]: OK

**Pull request merged**